engine/src/witness/btc/source.rs

```diff
@@ -40,22 +47,90 @@ where
 	) -> (BoxChainStream<'_, Self::Index, Self::Hash, Self::Data>, Self::Client) {
 		(
 			Box::pin(stream::unfold(
-				(self.client.clone(), None, make_periodic_tick(POLL_INTERVAL, true)),
-				|(client, last_block_hash_yielded, mut tick)| async move {
+				(
+					self.client.clone(),
+					Option::<BtcSourceState>::None,
+					make_periodic_tick(POLL_INTERVAL, true),
+				),
+				|(client, mut stream_state, mut tick)| async move {
 					loop {
+						// We don't want to wait for the tick if we have backfilling to do, so we do
+						// it here before awaiting the tick.
+						if let Some(state) = &stream_state {
+							if state.best_known_block_index > state.last_block_yielded_index {
+								tracing::debug!(
+									"Backfilling BTC source from index {} to {}",
+									state.last_block_yielded_index,
+									state.best_known_block_index,
+								);
+								let header = client
+									.header_at_index(
+										state.last_block_yielded_index.saturating_add(1),
+									)
+									.await;
+								return Some((
+									header,
+									(
+										client,
+										Some(BtcSourceState {
+											last_block_yielded_hash: header.hash,
+											last_block_yielded_index: header.index,
+											best_known_block_index: state.best_known_block_index,
+										}),
+										tick,
+									),
+								));
+							}
+						}
+
 						tick.tick().await;
 
 						let best_block_header = client.best_block_header().await;
-						if last_block_hash_yielded != Some(best_block_header.hash) {
+
+						let yield_new_best_header: bool = match &mut stream_state {
+							Some(state)
+								// We want to immediately yield the new best header if we've reorged on the same block
+								// or it's the next block we expect
+								if (state.last_block_yielded_index == best_block_header.height &&
+									state.last_block_yielded_hash != best_block_header.hash) ||
+									state.last_block_yielded_index.saturating_add(1) == best_block_header.height =>
+								true,
+							// If we don't yet have state (we're initialising), then we want to
+							// yield the new best header immediately
+							None => true,
+							// If we've progressed by more than one block, then we need to backfill
+							Some(state)
+								if state.last_block_yielded_index < best_block_header.height =>
+							{
+								// Update the state for the next iteration to backfill
+								if let Some(state) = stream_state.as_mut() {
+									state.best_known_block_index = best_block_header.height;
+								}
```

Comment:
We already have a mutable reference to `state`, why are we creating a new one? I'm actually surprised that this works, because it looks like we have two mut references to the same thing?